Family version redirects

maddiedierker opened the pull request:
Previously implemented in #27977, which I closed in order to implement a different approach.

[LP-28](https://codedotorg.atlassian.net/browse/LP-28)
[Spec](https://docs.google.com/document/d/1jYznJYlrfXQ-MJvmgfmZbVjjAM9-i297hoFTCqoYJ5w/edit?usp=sharing)

**Satisfies the following requirements in spec:**
If I go to a URL that is the “base name” of the script (e.g. /s/coursea), and...
- I have no existing progress in any version of that script: I should get redirected to the recommended one for my language. For example, if I’m in Spanish, /s/coursea should redirect me to /s/coursea-2017. If I’m in English or in an unsupported language (say Hindi), I should get redirected to /s/coursea-2018. 
- If I have existing progress in a newer version or I’m assigned to ONLY a newer version of the course than is recommended for my language, I should be redirected to the newest version I have progress in. EX: I have progress in Coursea-2018 in English, but my language is Spanish (2017 would be my recommended stable version), I should still go to coursea-2018 because that’s where I’ve been assigned or have progress.
- If I have existing progress in multiple versions of the script or am assigned to multiple versions of the script where some of the progress/assignment is in a newer version than is recommended for my language, I should still go to the latest version that I have progress in, but we should warn you that you have progress in other, older versions.

davidsbailey commented:
👁 👁 looking...

davidsbailey commented on dashboard/app/models/script.rb:
```diff
@@ -437,14 +437,18 @@ def self.get_from_cache(id_or_name, version_year: nil)
-    script_name = Script.latest_stable_version(family_name, version_year: version_year).try(:name)
+  def self.get_script_family_redirect(family_name, version_year: nil, user: nil, locale: 'en-US')
+    script_name = Script.latest_assigned_version(family_name, user).try(:name)
+    script_name ||= Script.latest_stable_version(family_name, version_year: version_year, locale: locale, fallback: true).try(:name)
```
I'm afraid we can't afford to do this either -- this lets each call to `ScriptsController#show` hit our database. We need this method to rely on the script cache. 

A problem you may run into is that `get_from_cache` calls `get_without_cache` which will `raise` if the script is not found, which will happen when you are looking for a family name. Here is what I would recommend:
* Create a whitelist of script family names in script_constants.rb. When asked to `show` one of these, look for all scripts with this family name rather than scripts with this name, and use locale and user to decide which to redirect to.
* when looking for scripts by family name, you will need to cache the result so that we don't hit the database every time someone goes to /s/coursea (for example). I would imagine that you'd define this `script_family_from_cache` and `script_family_without_cache`, which takes in a family name and returns an array of scripts belonging to that family (or maybe an array of script objects _redirecting_ to scripts in that family, or just script names, whatever makes sense).
* since this will be a new kind of cache entry, there will be some places in our setup code that these need to be populated. 
* we have some unit tests which count how many database requests we are making on certain controller routes once the script cache has been populated. those are great to run as you iterate to make sure that you aren't adding too many DB requests, so that we don't have to reason abstractly about cache hits and misses.

maddiedierker replied:
i just realized i never thanked you for this guidance 😄 thank you! this was really helpful

davidsbailey commented:
example of a query count test that needs to stay at zero, from the drone output:
```
test_should_get_/s/frozen#CachingTest (509.51s)
1460 | Wrong query count:
1461 | Script Load (0.0ms) SELECT `scripts`.* FROM `scripts` WHERE `scripts`.`family_name` = 'frozen' ORDER BY properties -> '$.version_year' DESC
1462 | app/models/script.rb:524:in `latest_stable_version'
1463 | app/models/script.rb:451:in `get_script_family_redirect'
1464 | app/controllers/scripts_controller.rb:128:in `set_script_for_show'
1465 | test/integration/caching_test.rb:17:in `block (2 levels) in <class:CachingTest>'
1466 | test/integration/caching_test.rb:16:in `block in <class:CachingTest>'
1467 | .
1468 | Expected: 0
1469 | Actual: 1
```

maddiedierker commented on lib/cdo/script_constants.rb:
```diff
@@ -165,6 +165,37 @@ module ScriptConstants
 
   DEFAULT_VERSION_YEAR = '2017'
 
+  # A whitelist of all family names for scripts.
```
I grabbed all of the family names from prod with the following query:

```
Script.all.map(&:family_name).uniq
```

davidsbailey commented on dashboard/app/models/script.rb:
```diff
@@ -436,6 +448,51 @@ def self.get_from_cache(id_or_name, version_year: nil)
+
+    script_family_cache.fetch(family_name) do
+      # Populate cache on miss.
+      script_cache[family_name] = Script.get_family_without_cache(family_name)
```
should be `script_family_cache[...] =` ? if not... weird things may happen

maddiedierker replied:
yes! i was originally using `script_cache` instead of `script_family_cache` and missed a spot. thank you!

davidsbailey commented on dashboard/app/models/script.rb:
```diff
@@ -436,6 +448,51 @@ def self.get_from_cache(id_or_name, version_year: nil)
+
+      is_supported = script.supported_locales&.include?(locale_str) || locale_str&.downcase&.start_with?('en')
+      if is_supported
+        latest_version = script if is_supported
```
❤️ the self-documenting variable name! questions/comments here:
1. remove redundant double-checking of `if is_supported` on this line
2. does the `start_with?('en')` capture just one locale, or several? it makes sense that a user in England would be able to use an American-English version of our course... just requesting that the intention be documented with a quick comment.

maddiedierker replied:
thank you! all english-speaking locales should be supported, so i'll add a comment around that

davidsbailey commented on dashboard/test/models/script_test.rb:
```diff
@@ -388,6 +402,46 @@ def populate_cache_and_disconnect_db
+
+    redirect_script = Script.get_script_family_redirect_for_user('csp', locale: 'it-IT')
+    assert_equal csp1_2018.name, redirect_script.redirect_to
+  end
```
Great test coverage!

davidsbailey commented on lib/cdo/script_constants.rb:
```diff
@@ -165,6 +165,37 @@ module ScriptConstants
+    CSP2 = 'csp2'.freeze,
+    CSP3 = 'csp3'.freeze,
+    CSP4 = 'csp4'.freeze,
+    CSP5 = 'csp5'.freeze,
```
nothing to do here right now... but this reminds me that the idea of "csp5" script family makes less sense over time, considering CSP 20-21 may have a totally different structure. ideally... we wouldn't even need these script families, and would rely on the course families, which control both (1) where you get redirected to from /courses/csd and (2) what shows up in the assignment selector. Just food for thought.

maddiedierker commented:
thanks, dave! i think i have addressed all of your comments, and i went ahead and made [LP-345](https://codedotorg.atlassian.net/browse/LP-345) to track updating calls to `Script.get_from_cache` with family names to use reference script versions instead
